### PR TITLE
[tune] fix pbt checkpoint_freq

### LINKED ISSUE #9036: [tune] Population-based training: broken when using keep_checkpoint_num
When using **population-based** training TUNE stops after some times throwing the following error:

`There are paused trials, but no more pending trials with sufficient resources.`

This is caused by not finding the latest checkpoint:

```
Failure # 1 (occurred at 2020-06-19_11-26-36)
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/ray/tune/ray_trial_executor.py", line 294, in start_trial
    self._start_trial(trial, checkpoint, train=train)
  File "/usr/local/lib/python3.6/dist-packages/ray/tune/ray_trial_executor.py", line 235, in _start_trial
    self.restore(trial, checkpoint)
  File "/usr/local/lib/python3.6/dist-packages/ray/tune/ray_trial_executor.py", line 673, in restore
    data_dict = TrainableUtil.pickle_checkpoint(value)
  File "/usr/local/lib/python3.6/dist-packages/ray/tune/trainable.py", line 62, in pickle_checkpoint
    checkpoint_dir = TrainableUtil.find_checkpoint_dir(checkpoint_path)
  File "/usr/local/lib/python3.6/dist-packages/ray/tune/trainable.py", line 87, in find_checkpoint_dir
    raise FileNotFoundError("Path does not exist", checkpoint_path)
FileNotFoundError: [Errno Path does not exist] /content/TRASH_TUNE_PBT_oversampling_mimic_densenet121/TUNE_Model_0_2020-06-19_11-24-215xncry9c/checkpoint_6/
```

The error appears to be somewhat random since it only appears after quite some iterations

The error can be reproduced in this [colab notebook](https://colab.research.google.com/drive/1-o896bEUm7DTvS24Do0btlqbSHre49MH?usp=sharing). **It is not a COLAB related issue since the same problem arises on our own server.**

@richardliaw Is this related to #8772 ?

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

The `CheckpointManager` keeps track of the `last_persistent_checkpoint`. However, this fails when the newest checkpoint is performing badly, and the `on_checkpoint()` method is called twice with the same checkpoint location. In that case, `last_persistent_checkpoint` points to a storage location that has been deleted.

With this PR, the old checkpoint directory is only deleted when it does not coincide with the new to-be-stored checkpoint.

This problem came up when resuming trials in PopulationBasedTraining. A previous PR #9470 tackled this problem, but introduced unwanted logic changes.

## Related issue number

Closes old PR #9470
Should fix #9036, #8441

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests

### CODE REVIEW

File: python/ray/tune/checkpoint_manager.py [hunk 1]
```diff
@@ -112,7 +112,8 @@ def on_checkpoint(self, checkpoint):
         self.newest_persistent_checkpoint = checkpoint
 
         # Remove the old checkpoint if it isn't one of the best ones.
-        if old_checkpoint.value and old_checkpoint not in self._membership:
+        if old_checkpoint.value and old_checkpoint not in self._membership \
+           and old_checkpoint.value != checkpoint.value:
```
Comment (richardliaw) [inline python/ray/tune/checkpoint_manager.py:116]: @krfricke is it possible to add a comprehensive comment (as you already did on the PR description) into the code? Just to explain what you're doing.

Comment (richardliaw) [reply python/ray/tune/checkpoint_manager.py:116]: Also, is there an invariant here? (i.e., don't process the checkpoint if the checkpoint is already tracked?)

**Comment by ujvl** [comment]
is it possible (cleaner?) to just return early at beginning of `on_checkpoint` if the checkpoint is the same as the newest? and also `logger.debug` when that happens

**Comment by krfricke** [comment]
I think in some cases the checkpoint directory might be the same though the checkpoint reults are not, e.g. when you specify your own checkpoint directory, or when you don't pass a step to `tune.make_checkpoint_dir`. In those cases it might be that our score improved and we would like to keep the checkpoint in `_membership` and `_best_checkpoints`. We would thus have to compare the results dicts, which we would have to do item by item. Thus I'd suggest to keep it like this.

**Comment by richardliaw** [comment]
Ah, there is (or I think there is) the invariant that every checkpoint should exist in a different checkpoint directory. 

I might not have made it explicit in the docs though.

**Comment by krfricke** [comment]
This doesn't seem to be the case with PBT, which seems to store checkpoints again after perturbation:
```
old = {'mean_accuracy': 31.200000000000003, 'done': False, 'timesteps_total': None, 'episodes_total': None,
      'training_iteration': 3, 'experiment_id': '759f9cf229964f02920ba4aed8e63a54', 'date': '2020-07-17_09-17-12',
      'timestamp': 1594970232, 'time_this_iter_s': 1.0967254638671875e-05, 'time_total_s': 3.218650817871094e-05,
      'pid': 8478, 'hostname': '...', 'node_ip': '192.168.0.28', 'config': {'a': 40, 'b': 0.8, 'c': 1},
      'time_since_restore': 1.0967254638671875e-05, 'timesteps_since_restore': 0, 'iterations_since_restore': 1,
      'trial_id': '835a1_00015', 'experiment_tag': '15_a=40,b=0.8'}
new = {'mean_accuracy': 31.200000000000003, 'done': False, 'timesteps_total': None, 'episodes_total': None,
      'training_iteration': 4, 'experiment_id': '759f9cf229964f02920ba4aed8e63a54', 'date': '2020-07-17_09-17-19',
      'timestamp': 1594970239, 'time_this_iter_s': 1.52587890625e-05, 'time_total_s': 4.744529724121094e-05,
      'pid': 8561, 'hostname': '...', 'node_ip': '192.168.0.28', 'config': {'a': 40, 'b': 0.8, 'c': 1},
      'time_since_restore': 1.52587890625e-05, 'timesteps_since_restore': 0, 'iterations_since_restore': 1,
      'trial_id': '835a1_00015', 'experiment_tag': '15_a=40,b=0.8@perturbed[c=0]'}
```

The difference here is in the `experiment_tag`, `training_iteration` and time-related attributes.

The affected metadata is irrelevant for the bookkeeping decisions (as long as `training_iteration` is monotonic, which should be the case).

If the same checkpoint directory is really not used multiple times, we can return early, I just pushed a commit for this. Should `tune.save_checkpoint()` raise a warning if someone tries to use the same checkpoint directory more than once?

**Review by richardliaw — APPROVED** [review]

**Pull request merged**